packages/axelar-soroban-std/src/shared_interfaces.rs

```diff
@@ -0,0 +1,398 @@
+use crate::ensure;
+use soroban_sdk::{
+    contractclient, symbol_short, Address, BytesN, ConversionError, Env, FromVal, IntoVal, String,
+    Topics, TryFromVal, Val, Vec,
+};
+
+#[contractclient(name = "OwnershipClient")]
+pub trait OwnershipInterface {
+    fn owner(env: &Env) -> Address;
+}
+
+#[contractclient(name = "UpgradeableClient")]
+pub trait UpgradeableInterface: OwnershipInterface {
+    /// Returns the current version of the contract.
+    fn version(env: &Env) -> String;
+
+    /// Upgrades the contract to a new WASM hash.
+    fn upgrade(env: &Env, new_wasm_hash: BytesN<32>);
+}
+
+pub trait MigratableInterface: UpgradeableInterface {
+    /// Data needed during the migration. Each contract can define its own data type.
+    type MigrationData: FromVal<Env, Val>;
+    /// Error type returned if the migration fails.
+    type Error: Into<soroban_sdk::Error>;
+
+    /// Migrates contract state after upgrading the contract code.
+    fn migrate(env: &Env, migration_data: Self::MigrationData) -> Result<(), Self::Error>;
+}
+
+/// Default implementation of the [OwnershipInterface] trait.
+pub fn owner(env: &Env) -> Address {
+    env.storage()
+        .instance()
+        .get(&storage::DataKey::SharedInterfaces_Owner)
+        .expect("owner must be set during contract construction")
+}
+
+/// Default implementation accompanying the [OwnershipInterface] trait. This should never be part of a contract interface,
+/// but allows contracts internally to set the owner.
+pub fn set_owner(env: &Env, owner: &Address) {
+    env.storage()
+        .instance()
+        .set(&storage::DataKey::SharedInterfaces_Owner, owner);
+}
+
+/// This function checks that the caller can authenticate as the owner of the contract,
+/// then upgrades the contract to a new WASM hash and prepares it for migration.
+pub fn upgrade<T: OwnershipInterface>(env: &Env, new_wasm_hash: BytesN<32>) {
+    T::owner(env).require_auth();
+
+    env.deployer().update_current_contract_wasm(new_wasm_hash);
+    start_migration(env);
+}
+
+/// This function checks that the caller can authenticate as the owner of the contract,
+/// then runs the custom_migration and finalizes the migration.
+/// An event is emitted when the migration, and with it the overall upgrade, is complete.
+/// Migration can only be run once, after the [upgrade] function has been called.
+pub fn migrate<T: UpgradeableInterface>(
+    env: &Env,
+    custom_migration: impl FnOnce(),
+) -> Result<(), MigrationError> {
+    T::owner(env).require_auth();
+
+    ensure_is_migrating(env)?;
+
+    custom_migration();
+    complete_migration(env);
+
+    emit_event_upgraded(
+        env,
+        UpgradedEvent {
+            version: T::version(env),
+        },
+    );
+
+    Ok(())
+}
+
+fn emit_event_upgraded(env: &Env, event: UpgradedEvent) {
```

Comment:
Maybe create an `Event` trait that has `topic`, `data` methods, and a default `emit` method that uses the other methods? So usage becomes something like:
```
UpgradedEvent {
    version: T::version(env),
}.emit(env);
```

Reply:
Actually, might be better to define the event to include a reference to Env
```
UpgradedEvent {
    env,
    version: T::version(env),
}.emit();
```
since constructing the topic/data might require `env` (e.g. hashing a payload to emit in the topic). Alternatively, passing `env` to `topic()` and `data()` from `emit`

Reply:
follow-up PR